README.md
<h1 align="center">URLBrute</h1>
<p align="center">Tool for brute-force directories/DNS on websites</p>
<p align="center"></p>

## Avaliable Modes
- Directory/File brute-force
- DNS brute-force

## Installation
### Building from Source
- Install dependencies
```bash
go get
```

- Build executable
```bash
go build
```

- If you want to install in the `$GOPATH/bin` folder
```bash
go install
```

### Using `go get`
```bash
go get github.com/ReddyyZ/urlbrute
```

- Updating
```bash
go get -u github.com/ReddyyZ/urlbrute
```

---

## Help
### Global
```bash
urlbrute --help

Tool for brute-force directories on websites

Usage:
   urlbrute {flags}
   urlbrute <command> {flags}

Commands:
   dir
   dns
   help                          displays usage informationn
   version                       displays version number

Flags:
   -h, --help                    displays usage information of the application or a command (default: false)
   -v, --version                 displays version number (default: false)
```

### `Dir`
```bash
urlbrute dir --help

Scan for diretories on website

Usage:
   urlbrute {flags}

Flags:
   -c, --code                    Filter results by status codes (default: 200,204,301,302,307,401,403)
   -h, --help                    displays usage information of the application or a command (default: false)
   -i, --interval                Interval between requests in ms (default: 300)
   -t, --timeout                 Request timeout (default: 10)
   -u, --url                     URL to scan
   -a, --useragent               Set User-Agent (default: urlbrute/1.0.0)
   -w, --wordlist                Wordlist to test
```

### `DNS`
```bash
urlbrute dns --help

Scan for subdomains

Usage:
   urlbrute {flags}

Flags:
   -s, --dnsserver               DNS Servers to resolve (default: 8.8.8.8,8.8.4.4)
   -d, --domain                  Domain to scan
   -h, --help                    displays usage information of the application or a command (default: false)
   -i, --interval                Interval between requests in ms (default: 300)
   -a, --ip                      Show IP address of domain (default: false)
   -q, --quiet                   Show only domain found (default: false)
   -r, --retry                   Retry times (default: 5)
   -v, --verbose                 Verbose level (default: false)
   -w, --wordlist                Wordlist to test
```

---

## Examples
- Default options
```bash
urlbrute dir -u google.com -w common.txt
```
<img src="img/Example.png" width="782" height="485">

- Showing 404 status code
```bash
urlbrute dir -u google.com -w common.txt --code 200,404
```
<img src="img/Example1.png" width="782" height="485">

- Changing user-agent
```bash
urlbrute dir -u google.com -w common.txt --useragent useragent/1.0
```
<img src="img/Example2.png" width="782" height="485">

- Searching for DNS
```bash
urlbrute dns -d google.com -w domain.txt --ip
```
<img src="img/Example3.png" width="782" height="485">

---

<h2 align="center">&lt;/&gt; by <a href="https://github.com/ReddyyZ">ReddyyZ</a></h2>

CHANGELOG.md
## URLBrute v1.0.0
- Added `dir` mode with basic directory brute-force

## URLBrute v1.0.1
- Bug fixes

## URLBrute v1.0.2
- Added `dns` brute-force mode
    - Multiple output modes
- Script structure changed

core/print.go
package core

import (
	"fmt"
	"net/http"

	"github.com/fatih/color"
)

var banner string = `
██╗   ██╗██████╗ ██╗     ██████╗ ██████╗ ██╗   ██╗████████╗███████╗
██║   ██║██╔══██╗██║     ██╔══██╗██╔══██╗██║   ██║╚══██╔══╝██╔════╝
██║   ██║██████╔╝██║     ██████╔╝██████╔╝██║   ██║   ██║   █████╗  
██║   ██║██╔══██╗██║     ██╔══██╗██╔══██╗██║   ██║   ██║   ██╔══╝  
╚██████╔╝██║  ██║███████╗██████╔╝██║  ██║╚██████╔╝   ██║   ███████╗
 ╚═════╝ ╚═╝  ╚═╝╚══════╝╚═════╝ ╚═╝  ╚═╝ ╚═════╝    ╚═╝   ╚══════╝
                                                                   `

func Banner() {
	red := color.New(color.FgRed)
	red.Println(banner)
}

//
//

func Success(data string) {
	green := color.New(color.FgGreen)
	green.Printf("[+]")
	fmt.Println(data)
}

func SuccessYellow(data string) {
	yellow := color.New(color.FgYellow)
	yellow.Printf("[+]")
	fmt.Println(data)
}

func Error(data string) {
	red := color.New(color.FgRed)
	red.Printf("[-]")
	fmt.Println(data)
}

func Info(data string) {
	blue := color.New(color.FgBlue)
	blue.Printf("[*]")
	fmt.Println(data)
}

//
//

func Find(slice []int, val int) (int, bool) {
	// Check if element existst in array

	for i, item := range slice {
		if item == val {
			return i, true
		}
	}
	return -1, false
}

//
//

func DNSFound(url string, domain string, ip string, verbose_level int) {
	if verbose_level == -1 {
		fmt.Println(domain)
	} else if verbose_level == 0 {
		fmt.Println(url)
	} else {
		yellowHi := color.New(color.FgHiYellow)

		fmt.Printf("%s ", url) // Print Dir
		yellowHi.Print("-> ")
		fmt.Println(ip)
	}

}

func DirFound(url string, statuscode int, codes []int) {
	if _, i := Find(codes, statuscode); !i {
		return
	}

	// Colors
	green := color.New(color.FgGreen)
	yellow := color.New(color.FgYellow)
	yellowHi := color.New(color.FgHiYellow)
	red := color.New(color.FgRed)
	blue := color.New(color.FgBlue)

	// Status codes
	azul := []int{100, 101, 102}                                                                                                                                                                                                        // Informational codes
	verde := []int{200, 201, 202, 203, 204, 205, 206, 207, 208, 226}                                                                                                                                                                    // Success codes
	amarelo := []int{300, 301, 302, 303, 304, 305, 307, 308}                                                                                                                                                                            // Redirection codes
	vermelho := []int{400, 401, 402, 403, 404, 405, 406, 407, 408, 409, 410, 411, 412, 413, 414, 415, 416, 417, 418, 421, 422, 423, 424, 426, 428, 429, 431, 444, 451, 499, 500, 501, 502, 503, 504, 505, 506, 507, 508, 510, 511, 599} // Error codes

	fmt.Printf("%s ", url) // Print Dir
	yellowHi.Print("-> ")

	// Print status code
	if _, i := Find(azul, statuscode); i {
		blue.Printf("%d %s\n", statuscode, http.StatusText(statuscode))

	} else if _, i := Find(verde, statuscode); i {
		green.Printf("%d %s\n", statuscode, http.StatusText(statuscode))

	} else if _, i := Find(amarelo, statuscode); i {
		yellow.Printf("%d %s\n", statuscode, http.StatusText(statuscode))

	} else if _, i := Find(vermelho, statuscode); i {
		red.Printf("%d %s\n", statuscode, http.StatusText(statuscode))

	} else {
		yellow.Printf("%d %s\n", statuscode, http.StatusText(statuscode))
	}
}


core/config.go
package core

import (
	"fmt"
	"strings"
)

type Config struct {
	Version     string
	Version_    string
	Description string
}

type Dir struct {
	Url        string
	Wordlist   string
	Useragent  string
	Code       string
	Timeout    int
	Interval   int
	Extensions string
}

type DNS struct {
	Domain    string
	Wordlist  string
	Interval  int
	Retry     int
	DNSServer []string
	Verbose   int
}

func NewConfig(version string, description string) *Config {
	return &Config{
		Version:     fmt.Sprintf("v%s", version),
		Version_:    version,
		Description: description,
	}
}

func NewDir(url string, wordlist string, useragent string, code string, timeout int, interval int, extensions string) *Dir {
	if !strings.Contains(url, "http") {
		url = fmt.Sprintf("http://%s", url)
	}

	return &Dir{
		Url:        url,
		Wordlist:   wordlist,
		Useragent:  useragent,
		Code:       code,
		Timeout:    timeout,
		Interval:   interval,
		Extensions: extensions,
	}
}

func NewDNS(domain string, wordlist string, interval int, retry int, dnsserver string, verbose int) *DNS {
	return &DNS{
		Domain:    domain,
		Wordlist:  wordlist,
		Interval:  interval,
		Retry:     retry,
		DNSServer: strings.Split(dnsserver, ","),
		Verbose:   verbose,
	}
}


core/request.go
package core

import (
	"fmt"
	"io"
	"net/http"
	"strconv"
	"strings"
	"sync"
	"time"
)

type Request struct {
	url        string
	useragent  string
	headers    string
	statuscode int
	body       io.ReadCloser
}

func sliceAtoi(sa []string) ([]int, error) {
	si := make([]int, 0, len(sa))
	for _, a := range sa {
		i, err := strconv.Atoi(a)
		if err != nil {
			return si, err
		}
		si = append(si, i)
	}
	return si, nil
}

func NewRequest(url string, dir *Dir, wg *sync.WaitGroup) *Request {
	codes, _ := sliceAtoi(strings.Split(dir.Code, ",")) // Create int array with codes to show on result

	client := &http.Client{
		Timeout: time.Duration(dir.Timeout) * time.Second,
	}

	req, err := http.NewRequest("GET", url, nil) // Create request
	if err != nil {
		Error(fmt.Sprintf("Error making request: %s", err.Error()))
		(*wg).Done()

		DirFound(strings.Replace(url, dir.Url, "", 1), 404, codes) // Print founded dir

		return &Request{
			url:        url,
			useragent:  dir.Useragent,
			statuscode: 404,
		}
	}

	req.Header.Set("User-Agent", dir.Useragent) // Set user-agent

	res, err := client.Do(req) // Make request

	if err != nil {
		Error(fmt.Sprintf("Error making request: %s", err.Error()))
		fmt.Println(err)
		(*wg).Done()

		DirFound(strings.Replace(url, dir.Url, "", 1), 404, codes) // Print founded dir
		return &Request{
			url:        url,
			useragent:  dir.Useragent,
			statuscode: 404,
		}
	}

	(*wg).Done()

	DirFound(strings.Replace(url, dir.Url, "", 1), res.StatusCode, codes) // Print founded dir
	return &Request{
		url:        url,
		useragent:  dir.Useragent,
		statuscode: res.StatusCode,
		body:       res.Body,
	}
}


brute/dirbrute.go
package brute

import (
	"bufio"
	"fmt"
	"io"
	"os"
	"strconv"
	"strings"
	"sync"
	"time"

	"github.com/ReddyyZ/urlbrute/core"
)

func sliceAtoi(sa []string) ([]int, error) {
	si := make([]int, 0, len(sa))
	for _, a := range sa {
		i, err := strconv.Atoi(a)
		if err != nil {
			return si, err
		}
		si = append(si, i)
	}
	return si, nil
}

func DirBrute(dir *core.Dir) {
	core.Banner()
	fmt.Printf(`
----------Target----------
Target: %s
Wordlist: %s
--------Req Config--------
User-Agent: %s
Timeout: %d
Interval: %dms
Codes: %s
--------------------------

`, dir.Url, dir.Wordlist, dir.Useragent, dir.Timeout, dir.Interval, dir.Code)

	// Variables declaration
	var wg sync.WaitGroup
	var line string
	results := make(chan *core.Request)

	// Open Wordlist
	file, err := os.Open(dir.Wordlist)
	if err != nil {
		core.Error(fmt.Sprintf("Error opening wordlist: %s", err.Error()))
		return
	}

	// Close on end
	defer file.Close()

	// Create I/O Reader
	reader := bufio.NewReader(file)

	core.Info("Scanning...\n")

	for {
		// Read line
		line, err = reader.ReadString('\n')
		line = strings.ReplaceAll(line, "\n", "")
		line = strings.ReplaceAll(line, "\r", "")

		// Handle read error
		if err != nil && err != io.EOF || line == "" {
			break
		}

		// Format URl
		url := fmt.Sprintf("%s/%s", dir.Url, line)

		// Make a new Request
		go func() {
			wg.Add(1)
			results <- core.NewRequest(url, dir, &wg)
		}()

		// Make requests with defined extensions
		if dir.Extensions != "no_extension" {
			time.Sleep(time.Duration(dir.Interval) * time.Millisecond)

			extensions := strings.Split(dir.Extensions, ",")

			for _, ext := range extensions {
				go func() {
					wg.Add(1)
					results <- core.NewRequest(fmt.Sprintf("%s.%s", url, ext), dir, &wg)
				}()

				time.Sleep(time.Duration(dir.Interval) * time.Millisecond)
			}
		}

		// Sleep
		time.Sleep(time.Duration(dir.Interval) * time.Millisecond)
	}

	go func() {
		wg.Wait()
		close(results)
	}()

	// for req := range results {
	// 	if _, i := Find(codes, req.statuscode); i {
	// 		DirFound(strings.Replace(req.url, dir.url, "", 1), req.statuscode)
	// 	}
	// }
}


brute/dnsbrute.go
package brute

import (
	"bufio"
	"fmt"
	"io"
	"os"
	"strings"
	"sync"
	"time"

	"github.com/ReddyyZ/urlbrute/core"
	"github.com/bogdanovich/dns_resolver"
)

func Resolve(url string, domain string, resolver *dns_resolver.DnsResolver, wg *sync.WaitGroup, verbose_level int) {
	if verbose_level == 2 {
		core.Info(fmt.Sprintf("Testing: %s", url))
	}

	// Resolve DNS
	ip, err := resolver.LookupHost(url)
	if err != nil {
		// if verbose_level == 2 {
		// 	fmt.Println("Erro:", url, "-", err.Error())
		// }
		wg.Done()
		return
	}

	// Print DNS founded
	core.DNSFound(url, domain, ip[0].String(), verbose_level)
	wg.Done()
}

func DNSBrute(dns *core.DNS) {
	if dns.Verbose != -1 {
		core.Banner()
		fmt.Printf(`
----------Target----------
Target: %s
Wordlist: %s
--------DNS Config--------
Interval: %dms
Retry: %d
DNSServer: %s
--------------------------

`, dns.Domain, dns.Wordlist, dns.Interval, dns.Retry, dns.DNSServer)
	}

	// Create resolver
	resolver := dns_resolver.New(dns.DNSServer)
	resolver.RetryTimes = dns.Retry

	// Declare variables
	var wg sync.WaitGroup
	var line string

	// Open wordlist
	fd, err := os.Open(dns.Wordlist)
	if err != nil {
		core.Error(fmt.Sprintf("Error opening wordlist: %s", err.Error()))
		return
	}
	defer fd.Close()

	// Create I/O Reader
	reader := bufio.NewReader(fd)

	if dns.Verbose != -1 {
		core.Info("Scanning...\n")
	}

	for {
		// Read a line
		line, err = reader.ReadString('\n')
		line = strings.ReplaceAll(line, "\n", "")
		line = strings.ReplaceAll(line, "\r", "")

		// Handle read error
		if err != nil && err != io.EOF || line == "" {
			break
		}

		// Format URL
		url := fmt.Sprintf("%s.%s", line, dns.Domain)

		// Resolve DNS func
		go func() {
			wg.Add(1)
			Resolve(url, line, resolver, &wg, dns.Verbose)
		}()

		// Sleep
		time.Sleep(time.Duration(dns.Interval) * time.Millisecond)
	}

	// Wait for threads to finish
	go func() {
		wg.Wait()
	}()
}


cli/urlbrute.go
package cli

import (
	"fmt"

	"github.com/ReddyyZ/urlbrute/brute"
	"github.com/ReddyyZ/urlbrute/core"
	"github.com/thatisuday/commando"
)

func Run(config *core.Config) {
	commando.
		SetExecutableName("urlbrute").
		SetVersion(config.Version).
		SetDescription(config.Description)

	commando.
		Register("dir").
		SetDescription("Scan for diretories on website").
		AddFlag("code,c", "Filter results by status codes", commando.String, "200,204,301,302,307,401,403").    // Optional
		AddFlag("timeout,t", "Request timeout", commando.Int, 10).                                              // Optional
		AddFlag("useragent,a", "Set User-Agent", commando.String, fmt.Sprintf("urlbrute/%s", config.Version_)). // Optional
		AddFlag("interval,i", "Interval between requests in ms", commando.Int, 300).                            // Optional
		AddFlag("extension,x", "Add extensions to end of each request", commando.String, "no_extension").       // Optional
		AddFlag("url,u", "URL to scan", commando.String, nil).                                                  // Required
		AddFlag("wordlist,w", "Wordlist to test", commando.String, nil).                                        // Required
		SetAction(func(args map[string]commando.ArgValue, flags map[string]commando.FlagValue) {
			url, _ := flags["url"].GetString()
			wordlist, _ := flags["wordlist"].GetString()
			code, _ := flags["code"].GetString()
			timeout, _ := flags["timeout"].GetInt()
			useragent, _ := flags["useragent"].GetString()
			interval, _ := flags["interval"].GetInt()
			extension, _ := flags["extension"].GetString()

			dir := core.NewDir(url, wordlist, useragent, code, timeout, interval, extension)
			brute.DirBrute(dir)
		})

	commando.
		Register("dns").
		SetDescription("Scan for subdomains").
		AddFlag("dnsserver,s", "DNS Servers to resolve", commando.String, "8.8.8.8,8.8.4.4").
		AddFlag("retry,r", "Retry times", commando.Int, 5).
		AddFlag("interval,i", "Interval between requests in ms", commando.Int, 300).
		AddFlag("quiet,q", "Show only domain found", commando.Bool, false).
		AddFlag("ip,a", "Show IP address of domain", commando.Bool, false).
		AddFlag("verbose,v", "Verbose level", commando.Bool, false).
		AddFlag("domain,d", "Domain to scan", commando.String, nil).     // Required
		AddFlag("wordlist,w", "Wordlist to test", commando.String, nil). // Required
		SetAction(func(args map[string]commando.ArgValue, flags map[string]commando.FlagValue) {
			domain, _ := flags["domain"].GetString()
			wordlist, _ := flags["wordlist"].GetString()
			interval, _ := flags["interval"].GetInt()
			retry, _ := flags["retry"].GetInt()
			dnsserver, _ := flags["dnsserver"].GetString()

			quiet, _ := flags["quiet"].GetBool()
			ip, _ := flags["ip"].GetBool()
			verbose, _ := flags["verbose"].GetBool()
			var verboseLevel int

			if quiet {
				verboseLevel = -1
			} else if ip {
				verboseLevel = 1
			} else if verbose {
				verboseLevel = 2
			} else {
				verboseLevel = 0
			}

			dns := core.NewDNS(domain, wordlist, interval, retry, dnsserver, verboseLevel)
			brute.DNSBrute(dns)
		})

	commando.Parse(nil)
}
